### DISCUSSION TITLE
Why does "colour.wavelength_to_XYZ" definition returns values outside expected range?

### DISCUSSION THREAD

**Discussion by LeCyberDucky** [colour-science/colour#746] [Q&A] [answered]
Hi, 

I'm trying to obtain RGB values of the visible spectrum. Therefore, I do this: 

```
import numpy as np
import colour
wavelength_range = np.arange(360, 830 + 0.5, 0.5)
rgb_spectrum = colour.XYZ_to_sRGB(colour.wavelength_to_XYZ(wavelength_range))
```

Based on this: 
https://colour.readthedocs.io/en/develop/generated/colour.wavelength_to_XYZ.html
https://colour.readthedocs.io/en/develop/generated/colour.XYZ_to_sRGB.html

I would expect the output to be in the range [0; 1]. From the above code, however, I get a maximum entry of 1.783001. Is there something wrong here, or am I simply missunderstanding the documentation? If the value is correct, I'm not sure how to interpret it. 

I have included the output from 
`$ python -c "import colour;colour.utilities.describe_environment()"`
below, in case that might be helpful. I hope somebody can help me with this :)

```
===============================================================================
*                                                                             *
*   Interpreter :                                                             *
*       python : 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 15 2019, 00:11:34) [MSC   *
*   v.1916 64 bit (AMD64)]                                                    *
*                                                                             *
*   colour-science.org :                                                      *
*       colour : 0.3.15                                                       *
*                                                                             *
*   Runtime :                                                                 *
*       imageio : 2.5.0                                                       *
*       matplotlib : 3.1.1                                                    *
*       networkx : 2.3                                                        *
*       numpy : 1.17.2                                                        *
*       pandas : 0.24.2                                                       *
*       scipy : 1.3.1                                                         *
*       six : 1.12.0                                                          *
*                                                                             *
===============================================================================
```

Edit: Having looked a bit more into this, I can see that not only do I get values greater than 1, I also get at negative minimum value of -9.270830e-08. 

Feeding the data into the XYZ_to_sRGB function results in minimum and maximum values of -11.956099 and 1.494786, respectively.

**Comment by Ron024** [comment]
It looks like you are trying to map the locus values (outer edge of the diagram) of the [CIE1931 space onto sRGB space](https://upload.wikimedia.org/wikipedia/commons/thumb/6/60/CIE1931xy_CIERGB.svg/325px-CIE1931xy_CIERGB.svg.png).  Most of these fall out side the sRGB space.  To interpolate to a value I believe they should be clipped to 0 or 1.

**Comment by LeCyberDucky** [comment]
Hmm, yes, that makes sense, thanks! 
I still have to wonder, though. The accepted range of values for `wavelength_to_XYZ` appears to be [360; 830] (based on the error message I get when trying values outside this range). And the documentation for the function indicates an output range of [0; 1] (if I'm reading it correctly). So, if my input values are in the correct range, it seems odd that the output would fall outside the expected range. 

So I suppose that I should simply clip the XYZ values to [0; 1] before converting to sRGB, right? 

I am basically trying to create spectral distribution plots just like what is created by the  function `colour.plotting.plot_single_sd()`. I just need the colour mapping however, since I'm trying to plot using Altair instead.

**Comment by KevinJW** [comment, marked as answer]
if all you want is the 'approximate colours' that are used in that function you could extract just the beginning of that function up to the point it generates the colours object (about half way through the function https://github.com/colour-science/colour/blob/3bebd7e903cbcbee959f56786c020335a1b06cf7/colour/plotting/colorimetry.py#L152)

As was mentioned above there is an adjustment being made to bring the colours in gamut for an sRGB device, as pretty much none of the wavelength buckets are likely to fall inside sRGB.

Kevin

**Comment by thomasmansencal** [comment]
Hi @LeCyberDucky,

To expand on @KevinJW and @Ron024, the reported domains and ranges in the documentation can be somehow misleading if you haven't read this portion of the documentation: https://colour.readthedocs.io/en/develop/basics.html#domain-range-scales.

You can think about them as indicative _soft_ limits. Colour is built to work on double-precision floating-point numbers. We promote everything to float64 by default so that we can manipulate HDRI imagery and values that are outside the common gamuts and even outside the spectral locus. This is critical when dealing with digital-still and motion-picture cameras. With that in mind, we try to never clip nor normalise anything, except in very rare instances, e.g. web colours.

I hope it makes sense!

**Comment by LeCyberDucky** [comment]
Yes, that makes sense. I hadn't seen that documentation page. Thank you very much.

**Comment by thomasmansencal** [comment]
You are welcome! :)
